src/plugins/kernel_install/utils.sh

```diff
@@ -554,23 +554,39 @@ function kernel_uninstall()
 {
   local reboot="$1"
   local target="$2"
-  local kernel_list_string="$3"
+  local kernel_list_string_or_regex="$3"
   local flag="$4"
   local force="$5"
   local prefix="$6"
+  local regex="$7"
   local update_grub=0
 
   cmd_manager "$flag" "sudo mkdir -p $REMOTE_KW_DEPLOY"
   cmd_manager "$flag" "sudo touch '$INSTALLED_KERNELS_PATH'"
 
-  kernel_list_string=$(printf '%s' "$kernel_list_string" | tr --delete ' ')
+  if [[ $regex == 1 ]]; then
+    regex_pattern=$kernel_list_string_or_regex
+    all_kernels=$(list_installed_kernels 'SILENT' 1 1 "$prefix")
+    IFS=, read -r -a available_kernels <<< "$all_kernels"
+    mapfile -t deduped_kernel_names < <(printf "%s\n" "${available_kernels[@]}" | sort -u)
+    kernel_names=()
+    for kernel in "${deduped_kernel_names[@]}"; do
+      # matching kernels to be uninstalled with the given regex
+      if [[ $kernel =~ $regex_pattern ]]; then
+        kernel_names+=("$kernel")
+      fi
+    done
```

Comment:
Hi, @lorenzoberts and @BrizaMel, and thank you guys so much for the change! You've really nailed in this first PR, in terms of following the project rules and practices :+1:

The change seems legit from pure reviewing (didn't test it yet), but I think that there is an easier and more robust approach to implement this. I think that we can leverage the fact that any execution of `kw deploy --uninstall` passes through this function (`kernel_uninstall`) to just alter the processing of the variable `kernel_list_string` (that contains the arguments for `--uninstall`) to **also** support regex arguments.

My idea was to generalize this excerpt of code to handle both types of arguments (full names and regexs) indistinguishably. For example, suppose `kernel_list_string='kernel1,kernel[2-4]'`, and the kernels installed on the target are `kernel1`, `kernel2`, and `kernel3`. When we get to the assertion

```bash
[[ "$kernel" =~ $full_name_or_regex ]]
```

`kernel1` will match with `kernel1`, and `kernel2` and `kernel3` will match with `kernel[2-4]`.

There is probably some complexity that I am overlooking (we would need to think about the boundaries of the regex, how to handle kernel names of the type `<some_kernel>+`, etc.), but IMHO I think that we can cleanly implement the proposed while keeping the feature interface untouched with this.

Feel free to ping me for any doubts related or not with this :smiley:

Reply:
Hi @davidbtadokoro and @rodrigosiqueira, thanks for the appointment! We believe that this solution may complicate the command's usability since users would need to escape all special characters. In our opinion, using a flag would be the best option as it already allows multiple patterns to be passed. In the case of this example, the input regex would be (kernel1|kernel[2-4]). Additionally, the implementation cost of adding a prefix to each element of the list may not be reflected in the impact on users. What's your opinion on this?

Reply:
Hi @lorenzoberts and @BrizaMel, and thank you for the update!

> Additionally, the implementation cost of adding a prefix to each element of the list may not be reflected in the impact on users. What's your opinion on this?

I'm not sure if you are saying that it would be a good approach or not, but, IMHO, I think it would have the best trade-off, as the change would be more self-contained, and we wouldn't need a new option.

Below, is a rough draft that I managed to implement which adds the regex support by just modifying `kernel_uninstall`, with the restriction that a regex arguments should be prefixed with `regex:`. (Note that I hardcoded the value`installed_kernels`, and added an `exit` statement to "lazy-test" this draft :sweat_smile:)

```bash
@@ -559,6 +559,10 @@ function kernel_uninstall()
   local force="$5"
   local prefix="$6"
   local update_grub=0
+  declare -A kernel_names
+  local installed_kernels
+  local kernel_name
+  local kernel_regex
 
   cmd_manager "$flag" "sudo mkdir -p $REMOTE_KW_DEPLOY"
   cmd_manager "$flag" "sudo touch '$INSTALLED_KERNELS_PATH'"
@@ -570,7 +574,29 @@ function kernel_uninstall()
     exit 22 #EINVAL
   fi
 
-  IFS=', ' read -r -a kernel_names <<< "$kernel_list_string"
+  installed_kernels='kernel1,kernel2,kernel3,kernel4'
+  # installed_kernels=$(list_installed_kernels 'SILENT' 1 1 "$prefix")
+  IFS=, read -r -a installed_kernels_array <<< "$installed_kernels"
+  IFS=', ' read -r -a kernel_name_or_regex_array <<< "$kernel_list_string"
+  for installed_kernel in "${installed_kernels_array[@]}"; do
+    for kernel_name_or_regex in "${kernel_name_or_regex_array[@]}"; do
+      # In case the element starts with `regex:` we use regex to match
+      if [[ "$kernel_name_or_regex" =~ ^regex: ]]; then
+        kernel_regex=$(printf '%s' "$kernel_name_or_regex" | cut --delimiter=':' --fields=2)
+        [[ "$installed_kernel" =~ ^$kernel_regex$ ]] && kernel_names["$installed_kernel"]=1
+      else
+        kernel_name="$kernel_name_or_regex"
+        [[ "$installed_kernel" == "$kernel_name" ]] && kernel_names["$installed_kernel"]=1
+      fi
+    done
+  done
+
+  for kernel_name in "${!kernel_names[@]}"; do
+    printf 'Removing: %s\n' "$kernel_name"
+  done
+
+  exit
+
   for kernel in "${kernel_names[@]}"; do
     cmd="sudo grep -q '$kernel' '$INSTALLED_KERNELS_PATH'"
     cmd_manager "$flag" "$cmd"
```

Below, is a screenshot of an example of execution in which the user passes one full name argument `kernel` and one regex argument `regex:kernel[2-4]`.

![2024-05-12-002055_867x501_scrot](https://github.com/kworkflow/kworkflow/assets/95055784/4f61320f-0732-4599-8b9c-cdfa91c133c3)

Notice that the function correctly interprets `regex:kernel[2-4]` as a regex argument and matches it with `kernel2`, `kernel3`, and `kernel4`, while not matching the full name `kernel` with any installed kernel.

Reply:
Hi @davidbtadokoro, the commit 0eebc3aafc8bb5cd4d40ad1832910cde32754397 was based on your draft. We adapted the tests we had done, and now everything seems to be fine, including the fix for the issue mentioned by @melissawen. Let us know what you think. 
ps: after all the remaining revision iterations and before merging we probably should squash this three commits into one.